src/Model_ELBDM/CPU_ELBDM/CPU_ELBDMSolver_HJ.cpp

```diff
@@ -0,0 +1,588 @@
+//                k_gap                 : Number of useless grids on each side in the k direction (k mya not be equal to z)
+//                s_In                  : Shared memory array to store the input data and the solutions at different times
+//                s_Flux                : Shared memory array to store the boundary fluxes
+//                s_HasWaveCounterpart  : Shared memory array to store where cells have wave counterpart
+//                FinalOut              : true --> store the updated data to g_Fluid_Out
+//                XYZ                   : 0 : Update the solution in the x direction
+//                                        3 : Update the solution in the y direction
+//                                        6 : Update the solution in the z direction
+//                                        --> This parameter is also used to determine the place to store the output fluxes
+//                MinDens               : Minimum allowed density
+//-------------------------------------------------------------------------------------------------------
+GPU_DEVICE
+void CUFLU_Advance(  real g_Fluid_In [][FLU_NIN  ][ CUBE(HYB_NXT) ],
+                     #ifdef GAMER_DEBUG
+                     real g_Fluid_Out[][FLU_NOUT ][ CUBE(PS2) ],
+                     #else
+                     real g_Fluid_Out[][FLU_NIN  ][ CUBE(PS2) ],
+                     #endif
+                     real g_Flux     [][9][NFLUX_TOTAL][ SQR(PS2) ],
+                     const bool g_IsCompletelyRefined[],
+                     const bool g_HasWaveCounterpart [][ CUBE(HYB_NXT) ],
+                     int NPatchGroup,
+                     const real dt, const real _dh, const real Eta, const bool StoreFlux,
+                     const uint j_gap, const uint k_gap,
+                     const bool FinalOut,
+                     const int XYZ, const real MinDens )
+{
+
+   const real dh           = real(1.0)/_dh;                   // grid spacing
+   const real Coeff1       = real(1.0) * dt /(dh * Eta);      // coefficient for continuity equation
+   const real Coeff2       = real(0.5) * dt /(dh * dh * Eta); // coefficient for HJ-equation
+
+   const uint size_j       = HYB_NXT - 2 * j_gap;             // number of y-columns to be updated
+   const uint size_k       = HYB_NXT - 2 * k_gap;             // number of z-columns to be updated
+   const uint NColumnTotal = size_j * size_k;                 // total number of data columns to be updated
+   const uint NThread      = CGPU_FLU_BLOCK_SIZE_X * CGPU_FLU_BLOCK_SIZE_Y;  // total number of threads in thread block
+
+
+// openmp pragma for the CPU solver
+#  ifndef __CUDACC__
+#  pragma omp parallel
+#  endif
+   {
+//    create memories for columns of various intermediate fields on stack or shared GPU memory
+      CGPU_SHARED real  s_In                 [CGPU_FLU_BLOCK_SIZE_Y][N_TIME_LEVELS + 1][FLU_NIN][HYB_NXT];
```

Comment:
@hyschive 
Perhaps. But I thought that shared memory did not benefit from memory coalescing. In my opinion, the global memory accesses in all the ELBDM solver probably have more potential for optimisation. Instead of tranposing the 3D arrays and then having coalesced 1D reads, we access the 3D block completely uncoalesced. I think it would be more useful to spend time on a effective 3D tranpose on GPU. In fact, Jowett already tried but did not get any speed gains. 
I will leave this code as is for now and write about what Jowett and I have tried in the manual.

Reply:
What shared memory really requires is to avoid [bank conflicts](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#shared-memory-5-x). But it may be difficult to achieve and, as you said, may not have a great impact on the overall performance if global memory access is the real bottleneck. So I agree with you and let's leave the code as is for now and just leave a comment like `//###OPTIMIZATION: change the order of different dimensions to [N_TIME_LEVELS + 1][FLU_NIN][CGPU_FLU_BLOCK_SIZE_Y][HYB_NXT]` for future reference.